pkg/operator/rbaccontroller/rbaccontroller.go

```diff
@@ -0,0 +1,90 @@
+package rbaccontroller
+
+import (
+	"context"
+	"strings"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	errors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Bug 1821771: We needed a way to tighten the cluster role binding when cluster
+	// is upgraded from 4.1
+
+	// openShiftManagedRoleBindingAnnotation is a cluster role binding annotation used
+	// by operator to manage it when set to true
+	openShiftManagedRoleBindingAnnotation = "openshiftapiserver.rbac.authorization.openshift.io/managed"
+)
+
+type RBACController struct {
+	kubeClient    kubernetes.Interface
+	eventRecorder events.Recorder
+}
+
+func NewRBACController(kubeClient kubernetes.Interface, informerFactory informers.SharedInformerFactory, recorder events.Recorder) factory.Controller {
+
+	informer := informerFactory.
+		Rbac().
+		V1().
+		ClusterRoleBindings().
+		Informer()
+
+	c := &RBACController{
+		kubeClient:    kubeClient,
+		eventRecorder: recorder,
+	}
+
+	return factory.New().
+		WithInformers(informer).
+		WithSync(c.sync).
+		ToController("RBACController", c.eventRecorder)
+}
+
+func (c *RBACController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	clusterRoleBindingNames := []string{"discovery", "system:basic-user", "system:discovery", "system:openshift:discovery", "cluster-status-binding"}
+
+	var errs []error
+
+	for _, clusterRoleBindingName := range clusterRoleBindingNames {
+		clusterRoleBinding, err := c.kubeClient.RbacV1().ClusterRoleBindings().Get(context.Background(), clusterRoleBindingName, metav1.GetOptions{})
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		reconcileClusterRoleBinding(clusterRoleBinding)
+
+		_, err = c.kubeClient.RbacV1().ClusterRoleBindings().Update(context.Background(), clusterRoleBinding, metav1.UpdateOptions{})
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+func reconcileClusterRoleBinding(clusterRoleBinding *rbacv1.ClusterRoleBinding) {
+	if clusterRoleBinding == nil {
+		return
+	}
+
+	val, ok := clusterRoleBinding.Annotations[openShiftManagedRoleBindingAnnotation]
+	if !ok || strings.Compare(val, "true") != 0 {
```

Comment:
why `strings.Compare`?

Reply:
was overthinking previously. switched to simple `==`